File: src/veritrail/image_geometry.py
from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass


MAX_IMAGE_WIDTH = 8192
MAX_IMAGE_HEIGHT = 8192
MAX_IMAGE_PIXELS = 16_777_216
MAX_IMAGE_DECODED_BYTES = 64 * 1024 * 1024

_PNG_SIGNATURE = b"\x89PNG\r\n\x1a\n"
_PNG_BIT_DEPTHS = {
    0: {1, 2, 4, 8, 16},
    2: {8, 16},
    3: {1, 2, 4, 8},
    4: {8, 16},
    6: {8, 16},
}
_JPEG_SOF_MARKERS = {
    0xC0,
    0xC1,
    0xC2,
    0xC3,
    0xC5,
    0xC6,
    0xC7,
    0xC9,
    0xCA,
    0xCB,
    0xCD,
    0xCE,
    0xCF,
}


class ImageGeometryError(ValueError):
    """Raised when compressed image metadata is malformed or exceeds the frozen budget."""


@dataclass(frozen=True)
class ImageGeometry:
    width: int
    height: int
    pixels: int
    decoded_bytes: int


def _bounded_geometry(width: int, height: int) -> ImageGeometry:
    if width <= 0 or height <= 0:
        raise ImageGeometryError("image dimensions must be positive")
    if width > MAX_IMAGE_WIDTH or height > MAX_IMAGE_HEIGHT:
        raise ImageGeometryError(
            f"image dimensions exceed {MAX_IMAGE_WIDTH}x{MAX_IMAGE_HEIGHT}"
        )
    pixels = width * height
    decoded_bytes = pixels * 4
    if pixels > MAX_IMAGE_PIXELS:
        raise ImageGeometryError(f"image pixel count exceeds {MAX_IMAGE_PIXELS}")
    if decoded_bytes > MAX_IMAGE_DECODED_BYTES:
        raise ImageGeometryError(
            f"estimated decoded image size exceeds {MAX_IMAGE_DECODED_BYTES} bytes"
        )
    return ImageGeometry(
        width=width,
        height=height,
        pixels=pixels,
        decoded_bytes=decoded_bytes,
    )


def _parse_png(content: bytes) -> ImageGeometry:
    if not content.startswith(_PNG_SIGNATURE):
        raise ImageGeometryError("image/png attachment is missing the PNG signature")

    offset = len(_PNG_SIGNATURE)
    geometry: ImageGeometry | None = None
    saw_idat = False
    saw_iend = False
    chunk_index = 0
    while offset < len(content):
        if len(content) - offset < 12:
            raise ImageGeometryError("PNG chunk header is truncated")
        length = struct.unpack_from(">I", content, offset)[0]
        chunk_type = content[offset + 4 : offset + 8]
        chunk_end = offset + 12 + length
        if chunk_end > len(content):
            raise ImageGeometryError("PNG chunk payload is truncated")
        if len(chunk_type) != 4 or any(
            not (65 <= byte <= 90 or 97 <= byte <= 122) for byte in chunk_type
        ):
            raise ImageGeometryError("PNG chunk type is malformed")
        payload = content[offset + 8 : offset + 8 + length]
        expected_crc = struct.unpack_from(">I", content, offset + 8 + length)[0]
        actual_crc = zlib.crc32(chunk_type + payload) & 0xFFFFFFFF
        if actual_crc != expected_crc:
            raise ImageGeometryError("PNG chunk CRC is invalid")

        if chunk_index == 0:
            if chunk_type != b"IHDR" or length != 13:
                raise ImageGeometryError("PNG must begin with a 13-byte IHDR chunk")
            width, height, bit_depth, color_type, compression, filtering, interlace = (
                struct.unpack(">IIBBBBB", payload)
            )
            if bit_depth not in _PNG_BIT_DEPTHS.get(color_type, set()):
                raise ImageGeometryError("PNG IHDR bit depth or color type is invalid")
            if compression != 0 or filtering != 0 or interlace not in {0, 1}:
                raise ImageGeometryError("PNG IHDR compression, filter, or interlace is invalid")
            geometry = _bounded_geometry(width, height)
        elif chunk_type == b"IHDR":
            raise ImageGeometryError("PNG contains more than one IHDR chunk")

        if chunk_type == b"IDAT":
            saw_idat = True
        if chunk_type == b"IEND":
            if length != 0 or chunk_end != len(content):
                raise ImageGeometryError("PNG IEND chunk is malformed or not final")
            saw_iend = True
            break

        offset = chunk_end
        chunk_index += 1

    if geometry is None or not saw_idat or not saw_iend:
        raise ImageGeometryError("PNG is missing required IHDR, IDAT, or IEND chunks")
    return geometry


def _parse_jpeg(content: bytes) -> ImageGeometry:
    if len(content) < 4 or content[:2] != b"\xff\xd8":
        raise ImageGeometryError("image/jpeg attachment is missing the JPEG SOI marker")

    offset = 2
    geometry: ImageGeometry | None = None
    in_scan = False
    while offset < len(content):
        if in_scan:
            marker_start = content.find(b"\xff", offset)
            if marker_start < 0 or marker_start + 1 >= len(content):
                raise ImageGeometryError("JPEG entropy stream is truncated")
            marker = content[marker_start + 1]
            if marker == 0x00:
                offset = marker_start + 2
                continue
            if marker == 0xFF:
                offset = marker_start + 1
                continue
            if 0xD0 <= marker <= 0xD7:
                offset = marker_start + 2
                continue
            offset = marker_start
            in_scan = False
            continue

        if content[offset] != 0xFF:
            raise ImageGeometryError("JPEG marker prefix is malformed")
        while offset < len(content) and content[offset] == 0xFF:
            offset += 1
        if offset >= len(content):
            raise ImageGeometryError("JPEG marker is truncated")
        marker = content[offset]
        offset += 1
        if marker == 0x00:
            raise ImageGeometryError("JPEG contains an escaped byte outside scan data")
        if marker == 0xD9:
            if geometry is None or offset != len(content):
                raise ImageGeometryError("JPEG EOI marker is missing metadata or is not final")
            return geometry
        if marker == 0xD8:
            raise ImageGeometryError("JPEG contains an unexpected SOI marker")
        if marker == 0x01 or 0xD0 <= marker <= 0xD7:
            continue
        if len(content) - offset < 2:
            raise ImageGeometryError("JPEG segment length is truncated")
        segment_length = struct.unpack_from(">H", content, offset)[0]
        if segment_length < 2 or offset + segment_length > len(content):
            raise ImageGeometryError("JPEG segment is malformed or truncated")
        payload_start = offset + 2
        payload_end = offset + segment_length

        if marker in _JPEG_SOF_MARKERS:
            if segment_length < 8:
                raise ImageGeometryError("JPEG SOF segment is too short")
            precision = content[payload_start]
            height, width = struct.unpack_from(">HH", content, payload_start + 1)
            components = content[payload_start + 5]
            if precision not in {8, 12} or components == 0:
                raise ImageGeometryError("JPEG SOF precision or component count is invalid")
            if segment_length != 8 + 3 * components:
                raise ImageGeometryError("JPEG SOF component metadata is malformed")
            next_geometry = _bounded_geometry(width, height)
            if geometry is not None and geometry != next_geometry:
                raise ImageGeometryError("JPEG contains conflicting SOF dimensions")
            geometry = next_geometry

        offset = payload_end
        if marker == 0xDA:
            if geometry is None:
                raise ImageGeometryError("JPEG scan begins before SOF dimensions")
            in_scan = True

    raise ImageGeometryError("JPEG is missing a final EOI marker")


def parse_image_geometry(content: bytes, media_type: str) -> ImageGeometry:
    if not isinstance(content, bytes):
        raise ImageGeometryError("image content must be bytes")
    if media_type == "image/png":
        return _parse_png(content)
    if media_type == "image/jpeg":
        return _parse_jpeg(content)
    raise ImageGeometryError(f"unsupported image media type: {media_type}")

File: web/src/domain/image-geometry.ts
export const MAX_IMAGE_WIDTH = 8192
export const MAX_IMAGE_HEIGHT = 8192
export const MAX_IMAGE_PIXELS = 16_777_216
export const MAX_IMAGE_DECODED_BYTES = 64 * 1024 * 1024

const PNG_SIGNATURE = [137, 80, 78, 71, 13, 10, 26, 10]
const PNG_BIT_DEPTHS = new Map<number, Set<number>>([
  [0, new Set([1, 2, 4, 8, 16])],
  [2, new Set([8, 16])],
  [3, new Set([1, 2, 4, 8])],
  [4, new Set([8, 16])],
  [6, new Set([8, 16])],
])
const JPEG_SOF_MARKERS = new Set([
  0xc0, 0xc1, 0xc2, 0xc3, 0xc5, 0xc6, 0xc7, 0xc9, 0xca, 0xcb, 0xcd, 0xce, 0xcf,
])
const CRC32_TABLE = (() => {
  const table = new Uint32Array(256)
  for (let index = 0; index < table.length; index += 1) {
    let value = index
    for (let bit = 0; bit < 8; bit += 1) {
      value = (value >>> 1) ^ (value & 1 ? 0xedb88320 : 0)
    }
    table[index] = value >>> 0
  }
  return table
})()

export type ImageGeometryErrorCode = 'INVALID_IMAGE' | 'IMAGE_GEOMETRY_LIMIT'

export class ImageGeometryError extends Error {
  readonly code: ImageGeometryErrorCode

  constructor(code: ImageGeometryErrorCode, message: string) {
    super(message)
    this.name = 'ImageGeometryError'
    this.code = code
  }
}

export interface ImageGeometry {
  width: number
  height: number
  pixels: number
  decodedBytes: number
}

function invalid(message: string): never {
  throw new ImageGeometryError('INVALID_IMAGE', message)
}

function readU16(bytes: Uint8Array, offset: number): number {
  return bytes[offset]! * 0x100 + bytes[offset + 1]!
}

function readU32(bytes: Uint8Array, offset: number): number {
  return (
    bytes[offset]! * 0x1000000 +
    bytes[offset + 1]! * 0x10000 +
    bytes[offset + 2]! * 0x100 +
    bytes[offset + 3]!
  )
}

function boundedGeometry(width: number, height: number): ImageGeometry {
  if (width <= 0 || height <= 0) invalid('图片宽高必须为正数。')
  if (width > MAX_IMAGE_WIDTH || height > MAX_IMAGE_HEIGHT) {
    throw new ImageGeometryError(
      'IMAGE_GEOMETRY_LIMIT',
      `图片宽高超过 ${MAX_IMAGE_WIDTH}×${MAX_IMAGE_HEIGHT} 的固定上限。`,
    )
  }
  const pixels = width * height
  const decodedBytes = pixels * 4
  if (pixels > MAX_IMAGE_PIXELS || decodedBytes > MAX_IMAGE_DECODED_BYTES) {
    throw new ImageGeometryError(
      'IMAGE_GEOMETRY_LIMIT',
      '图片总像素或预计解码内存超过固定上限。',
    )
  }
  return { width, height, pixels, decodedBytes }
}

function crc32(bytes: Uint8Array, start: number, end: number): number {
  let crc = 0xffffffff
  for (let index = start; index < end; index += 1) {
    crc = (crc >>> 8) ^ CRC32_TABLE[(crc ^ bytes[index]!) & 0xff]!
  }
  return (crc ^ 0xffffffff) >>> 0
}

function parsePng(bytes: Uint8Array): ImageGeometry {
  if (
    bytes.length < PNG_SIGNATURE.length ||
    PNG_SIGNATURE.some((value, index) => bytes[index] !== value)
  ) {
    invalid('PNG 签名无效。')
  }

  let offset = PNG_SIGNATURE.length
  let geometry: ImageGeometry | undefined
  let sawIdat = false
  let sawIend = false
  let chunkIndex = 0
  while (offset < bytes.length) {
    if (bytes.length - offset < 12) invalid('PNG 数据块头被截断。')
    const length = readU32(bytes, offset)
    const typeStart = offset + 4
    const payloadStart = offset + 8
    const payloadEnd = payloadStart + length
    const chunkEnd = payloadEnd + 4
    if (chunkEnd > bytes.length) invalid('PNG 数据块负载被截断。')
    const typeBytes = bytes.subarray(typeStart, payloadStart)
    if (
      typeBytes.length !== 4 ||
      [...typeBytes].some((value) => !((value >= 65 && value <= 90) || (value >= 97 && value <= 122)))
    ) {
      invalid('PNG 数据块类型无效。')
    }
    const type = String.fromCharCode(...typeBytes)
    if (crc32(bytes, typeStart, payloadEnd) !== readU32(bytes, payloadEnd)) {
      invalid('PNG 数据块 CRC 无效。')
    }

    if (chunkIndex === 0) {
      if (type !== 'IHDR' || length !== 13) invalid('PNG 必须以 13 字节 IHDR 开始。')
      const width = readU32(bytes, payloadStart)
      const height = readU32(bytes, payloadStart + 4)
      const bitDepth = bytes[payloadStart + 8]!
      const colorType = bytes[payloadStart + 9]!
      const compression = bytes[payloadStart + 10]!
      const filtering = bytes[payloadStart + 11]!
      const interlace = bytes[payloadStart + 12]!
      if (!PNG_BIT_DEPTHS.get(colorType)?.has(bitDepth)) {
        invalid('PNG IHDR 位深或颜色类型无效。')
      }
      if (compression !== 0 || filtering !== 0 || (interlace !== 0 && interlace !== 1)) {
        invalid('PNG IHDR 压缩、过滤或隔行参数无效。')
      }
      geometry = boundedGeometry(width, height)
    } else if (type === 'IHDR') {
      invalid('PNG 包含重复 IHDR。')
    }

    if (type === 'IDAT') sawIdat = true
    if (type === 'IEND') {
      if (length !== 0 || chunkEnd !== bytes.length) invalid('PNG IEND 无效或不是末块。')
      sawIend = true
      break
    }
    offset = chunkEnd
    chunkIndex += 1
  }
  if (!geometry || !sawIdat || !sawIend) invalid('PNG 缺少 IHDR、IDAT 或 IEND。')
  return geometry
}

function parseJpeg(bytes: Uint8Array): ImageGeometry {
  if (bytes.length < 4 || bytes[0] !== 0xff || bytes[1] !== 0xd8) {
    invalid('JPEG SOI 标记无效。')
  }

  let offset = 2
  let geometry: ImageGeometry | undefined
  let inScan = false
  while (offset < bytes.length) {
    if (inScan) {
      const markerStart = bytes.indexOf(0xff, offset)
      if (markerStart < 0 || markerStart + 1 >= bytes.length) invalid('JPEG 扫描流被截断。')
      const marker = bytes[markerStart + 1]!
      if (marker === 0x00) {
        offset = markerStart + 2
        continue
      }
      if (marker === 0xff) {
        offset = markerStart + 1
        continue
      }
      if (marker >= 0xd0 && marker <= 0xd7) {
        offset = markerStart + 2
        continue
      }
      offset = markerStart
      inScan = false
      continue
    }

    if (bytes[offset] !== 0xff) invalid('JPEG 标记前缀无效。')
    while (offset < bytes.length && bytes[offset] === 0xff) offset += 1
    if (offset >= bytes.length) invalid('JPEG 标记被截断。')
    const marker = bytes[offset]!
    offset += 1
    if (marker === 0x00) invalid('JPEG 扫描外出现转义字节。')
    if (marker === 0xd9) {
      if (!geometry || offset !== bytes.length) invalid('JPEG EOI 缺少元数据或不是末标记。')
      return geometry
    }
    if (marker === 0xd8) invalid('JPEG 包含意外 SOI。')
    if (marker === 0x01 || (marker >= 0xd0 && marker <= 0xd7)) continue
    if (bytes.length - offset < 2) invalid('JPEG 段长度被截断。')
    const segmentLength = readU16(bytes, offset)
    if (segmentLength < 2 || offset + segmentLength > bytes.length) {
      invalid('JPEG 段无效或被截断。')
    }
    const payloadStart = offset + 2
    const payloadEnd = offset + segmentLength

    if (JPEG_SOF_MARKERS.has(marker)) {
      if (segmentLength < 8) invalid('JPEG SOF 段过短。')
      const precision = bytes[payloadStart]!
      const height = readU16(bytes, payloadStart + 1)
      const width = readU16(bytes, payloadStart + 3)
      const components = bytes[payloadStart + 5]!
      if ((precision !== 8 && precision !== 12) || components === 0) {
        invalid('JPEG SOF 精度或分量数无效。')
      }
      if (segmentLength !== 8 + 3 * components) invalid('JPEG SOF 分量元数据无效。')
      const nextGeometry = boundedGeometry(width, height)
      if (
        geometry &&
        (geometry.width !== nextGeometry.width || geometry.height !== nextGeometry.height)
      ) {
        invalid('JPEG 包含冲突的 SOF 宽高。')
      }
      geometry = nextGeometry
    }

    offset = payloadEnd
    if (marker === 0xda) {
      if (!geometry) invalid('JPEG 在 SOF 宽高之前开始扫描。')
      inScan = true
    }
  }
  return invalid('JPEG 缺少最终 EOI。')
}

export function parseImageGeometry(bytes: Uint8Array, mediaType: string): ImageGeometry {
  if (mediaType === 'image/png') return parsePng(bytes)
  if (mediaType === 'image/jpeg') return parseJpeg(bytes)
  return invalid(`不支持的图片类型：${mediaType}`)
}

export async function validateImageGeometry(blob: Blob, mediaType: string): Promise<ImageGeometry> {
  return parseImageGeometry(new Uint8Array(await blob.arrayBuffer()), mediaType)
}
